crates/hulk_imagine/src/serializer.rs

```diff
@@ -0,0 +1,225 @@
+use std::collections::HashMap;
+
+use rmp_serde::encode::Error;
+use serde::{
+    ser::{self, Impossible, SerializeStruct},
+    Serialize,
+};
+
+#[derive(Default)]
+pub struct Serializer {
+    map: HashMap<String, Vec<u8>>,
+    stack: Vec<&'static str>,
+}
+
+impl Serializer {
+    pub fn finish(self) -> HashMap<String, Vec<u8>> {
+        self.map
+    }
+}
+
+impl<'a> ser::Serializer for &'a mut Serializer {
+    type Ok = ();
+    type Error = Error;
+
+    type SerializeSeq = Impossible<(), Error>;
+    type SerializeTuple = Impossible<(), Error>;
+    type SerializeTupleStruct = Impossible<(), Error>;
+    type SerializeTupleVariant = Impossible<(), Error>;
+    type SerializeMap = Impossible<(), Error>;
+    type SerializeStruct = Compound<'a>;
+    type SerializeStructVariant = Impossible<(), Error>;
+
+    fn serialize_bool(self, _v: bool) -> Result<Self::Ok, Self::Error> {
+        panic!("did not expect bool")
```

Comment:
better not use `panic` but rather proper errors here, all these functions already return a `Result`

Reply:
done